mesonpy/_editable.py

```diff
@@ -316,6 +328,44 @@ def rebuild(self) -> Node:
             install_plan = json.load(f)
         return collect(install_plan)
 
+    def _path_hook(self, path: str) -> MesonpyPathFinder:
+        if os.altsep:
+            path.replace(os.altsep, os.sep)
+        path, _, key = path.rpartition(os.sep)
+        if path == __file__:
+            tree = self._rebuild()
+            node = tree.get(tuple(key.split('.')))
+            if isinstance(node, Node):
+                return MesonpyPathFinder(node)
+        raise ImportError
+
+
+class MesonpyPathFinder(importlib.abc.PathEntryFinder):
+    def __init__(self, tree: Node):
+        self._tree = tree
+
+    def find_spec(self, fullname: str, target: Optional[ModuleType] = None) -> Optional[importlib.machinery.ModuleSpec]:
+        return find_spec(fullname, self._tree)
+
+    def iter_modules(self, prefix: str) -> Iterator[Tuple[str, bool]]:
+        yielded = set()
+        for name, node in self._tree.items():
+            ispkg = False
+            modname = inspect.getmodulename(name)
+            if modname == '__init__' or modname in yielded:
+                continue
+            if isinstance(node, Node):
+                modname = name
+                for _, suffix in LOADERS:
+                    src = node.get('__init__' + suffix)
+                    if isinstance(src, str):
+                        ispkg = True
+            if modname and '.' not in modname:
+                yielded.add(modname)
+                yield prefix + modname, ispkg
```

Comment:
I've found one issue where `pkgutil.walk_packages` returns directories with files in it that get installed with `install_subdir` but don't have an `__init__.py`. According to the docs, only modules should be returned. 

It looks like the condition here should include an `is_dir` part, and if it's a directory then `ispkg` must be True for this to be yielded. Although not sure if that will mess with namespace packages; this is handled in the `find_spec` function higher up, but not in this `iter_modules`.

I ran into this in both NumPy and SciPy, where the `test_public_api.py::test_all_modules_are_expected` tests turn up a few of these non-package directories.

Reply:
You're right. My understanding was that namespace packages should be returned too. I even added  test to check that this was the case. However, I've just tested and they are indeed not returned. Let me fix this.

Reply:
Thanks for the review @rgommers. The issue with namespace packages should be fixed now.

Reply:
The test failures are due to mesonbuild/meson#12920

Reply:
Thanks! I can confirm that it works now with both numpy and scipy, and from some more debug printing it seems to function exactly as expected.

Test failures with Meson master shouldn't be blocking here, so let's get this in.